tests/python/relax/test_transform_convert_dataflow.py

```diff
@@ -0,0 +1,493 @@
+            _ = R.print(format="Hi mom!")
+            a = R.multiply(v, v)
+            b = R.add(a, a)
+            c = R.subtract(b, a)
+            d = R.add(c, c)
+            return d
+
+    @I.ir_module
+    class Expected:
+        @R.function(pure=False)
+        def main(x: R.Tensor, y: R.Tensor) -> R.Tensor:
+            with R.dataflow():
+                z = R.add(x, y)
+                w = R.multiply(z, y)
+                v = R.add(w, x)
+                R.output(v)
+            _ = R.print(format="Hi mom!")
+            with R.dataflow():
+                a = R.multiply(v, v)
+                b = R.add(a, a)
+                c = R.subtract(b, a)
+                d = R.add(c, c)
+                R.output(d)
+            return d
+
+
+class TestExtractInsideBranches(ExtractCompare):
+    @I.ir_module
+    class Before:
+        @R.function(pure=False)
+        def main(x: R.Tensor, y: R.Tensor) -> R.Tensor:
+            z = R.add(x, y)
+            w = R.multiply(z, y)
+            v = R.add(w, x)
+            if R.const(True):
+                q = R.multiply(v, v)
+                a = R.add(q, q)
+                b = R.multiply(a, a)
+            else:
+                q = R.add(v, v)
+                a = R.multiply(q, q)
+                b = R.add(a, a)
+            c = R.multiply(b, b)
+            d = R.add(c, c)
+            e = R.multiply(d, d)
+            return e
+
+    @I.ir_module
+    class Expected:
+        @R.function(pure=False)
+        def main(x: R.Tensor, y: R.Tensor) -> R.Tensor:
+            with R.dataflow():
+                z = R.add(x, y)
+                w = R.multiply(z, y)
+                v = R.add(w, x)
+                R.output(v)
+
+            if R.const(True):
+                with R.dataflow():
+                    q = R.multiply(v, v)
+                    a = R.add(q, q)
+                    b = R.multiply(a, a)
+                    R.output(b)
+                # weird but the parser requires this construct
+                c = b
+            else:
+                with R.dataflow():
+                    q = R.add(v, v)
+                    a = R.multiply(q, q)
+                    b = R.add(a, a)
+                    R.output(b)
+                c = b
+            with R.dataflow():
+                d = R.multiply(c, c)
+                e = R.add(d, d)
+                f = R.multiply(e, e)
+                R.output(f)
+            return f
+
+
+class TestTreatNonCallAsPure(ExtractCompare):
+    @I.ir_module
+    class Before:
+        @R.function
+        def tuples_and_const(x: R.Tensor, y: R.Tensor) -> R.Tensor:
+            t1 = (x, y, x)
+            t2 = (y, y, x)
+            c = R.const([1, 2, 3], dtype="int32")
+            return c
+
+        @R.function
+        def shapes() -> R.Shape:
+            s1 = R.shape((1, 2, 3))
+            s2 = R.shape((4, 5, 6))
+            s3 = R.shape((7, 8, 9))
+            return s3
+
+        @R.function
+        def prim_values():
```

Comment:
Here's a bug I encountered, probably with the parser: Putting `-> R.Prim` as the return would segfault. Maybe it expects `-> R.Prim()`? Not sure why it would segfault though.

Reply:
It appears that it is not permitted to leave both optional fields empty. I guess it's not such a serious bug if it's not permitted anyway but it would be better to have a nicer error than a segfault.

Reply:
Whoa, that's an impressive segfault.  It looks like the sequence of events is as follows:

1. We hit [this TypeError](https://github.com/apache/tvm/blob/unity/python/tvm/script/parser/relax/entry.py#L452), which gives a diagnostic that the `R.Prim` needs to have one of the two argument types.
2. The TypeError is caught [here](https://github.com/apache/tvm/blob/unity/python/tvm/script/parser/relax/parser.py#L102), where it is passed to the diagnostic context to pretty-print and re-throw.
3. After showing the error message, [this conditional](https://github.com/apache/tvm/blob/unity/src/ir/diagnostic.cc#L129) resets the diagnostic renderer and raises an exception.
4. The DiagnosticError is caught [here](https://github.com/apache/tvm/blob/unity/python/tvm/script/parser/relax/parser.py#L111), where it is passed to the diagnostic context to pretty-print and re-throw.
5. When reaching [here](https://github.com/apache/tvm/blob/unity/src/ir/diagnostic.cc#L118), `(*this)->renderer` has been replaced with a default constructed instance.  The default-constructed instance from step (3) has a nullptr for the renderer, which causes the segfault.

It looks like there's a lot of `except DiagnosticError: raise` scattered throughout the parser to prevent this from occurring (by preventing step (4) from occurring), but these `eval_struct_info` and `eval_struct_info_proxy` don't have it.  That seems like a rather fragile way to avoid some pretty opaque errors, so I'm wondering if we want to change that in the long term.

Reply:
I think a more stable way to have the pretty-printing from TVMScript would be to remove all of the try/catch that only exist to re-throw as DiagnosticError, and instead allow the exception to bubble up to the entry point.  Catching the exception at the entry point, the `exc.__traceback__` can then be inspected.  The TVMScript location of the error is then the lowest stack frame that contains a variable of type `doc.AST`, and that object can be provided to `Parser.report_error` as the location of the error in the TVMScript.

This would avoid requiring every node to have its own try/catch/re-throw, and avoid the segfault that we ran into here.

@junrushao Since you've spent more time on the TVMScript parser than I have, what are your thoughts on this idea?